### DISCUSSION TITLE
Boolean variables don't work in x-bind

### DISCUSSION THREAD

**Discussion by maxsite** [alpinejs/alpine#2422] [1. Help] [answered]
Sample code:

```
    <div x-data="{tab: 1}">
        <details :open="tab == 1" @click="tab = 1">
            <summary>tab1</summary>
            <div>text</div>
        </details>

        <details :open="tab == 2" @click="tab = 2">
            <summary>tab2</summary>
            <div>text</div>
        </details>

        <details :open="tab == 3" @click="tab = 3">
            <summary>tab3</summary>
            <div>text</div>
        </details>
        
       <hr>
        <div x-text="tab"></div> <!-- for control -->
    </div>
```

Work in Alpine 2:
https://codepen.io/maxsite/pen/VwMZXKy

Incorrect in Alpine 3:
https://codepen.io/maxsite/pen/MWEgVJa

How to fix?

**Comment by SimoTod** [comment]
Remove :open, you don't need it browser already toggle open for you.
V3 is faster and it sets open to true before the browser handles it so your browser will remove the open attribute. (p.s. You don't need it in v2 either)

**Reply by maxsite** [reply]
This is necessary to automatically close other blocks.

**Reply by SimoTod** [reply]
Then use the prevent modifier so the browser doesn't toggle open since you are setting it already.

**Reply by SimoTod** [reply]
`@click.prevent="tab = 1"`

**Reply by maxsite** [reply]
It doesn't work that way. It is not possible to close an open block with prevent.

**Reply by SimoTod** [reply]
Prevent works, see https://codepen.io/SimoTod/pen/VwMZGwq
it's not the click that close the other tabs, clicks sets a value to a variable, :open closes the tabs and being reactive it updates every time open changes so the modifier won't really affect it.

The solution you posted below works because you are delaying the click code for 1ms, given time to the browser to do its stuff but it's not super clean, imo

**Reply by maxsite** [reply]
It is not possible to close an open block this way. Check out this example for the Alpine 2.

**Reply by SimoTod** [reply]
I thought we were talking about v3

**Reply by maxsite** [reply]
In the new version alpine works my code with debounce.

**Reply by SimoTod** [reply]
Yeah, but it's an odd hack, anything time based where the value is just a guess always sounds wrong to me. 
How can you be sure that 1ms is enough for all browsers and any conditions (busy CPU, energy saving mode, etc)? It probably is but it's an unnecessary risk if you know what the issue is. 

Btw just checked v2 and it does work with prevent so I'm not sure why you said it didn't. 
v2: https://codepen.io/SimoTod/pen/jOGNvam
v3:
https://codepen.io/SimoTod/pen/VwMZGwq

Obviously you can use whatever solution you like, glad you got it sorted.

**Reply by maxsite** [reply]
I have no idea why this works. :-) I need to transfer the code from alpine 2 to alpine 3. Why it does not work in its original form is a mystery to me. Because of this, you will have to use a hack.

**Reply by SimoTod** [reply]
Yeah but i posted you 2 links showing that prevent does work and you keep insisting that it doesn't. 

The reason is: 
1) When you click on a detail tag, the browser automatically invert the open attribute (no alpine, standard behaviour) 
2) if you set a `@click` attribute, or an old style onclick, the code in that attribute runs before the browser changes open. 
3) In v2, any update causes a refresh for all components in page and it's delayed naturally to give the code time to bundle multiple changes together and minimise the updates. In the original example, the browser changes open from false to true and Alpine sets it to true again after. 
4) in v3, only changed things are updated and they happen without delays. Your code sets open to true and then the browser sees it as true and revert it to false. 
5) prevent means "run this code but don't do what the browser would normally do" so you would just delegate to Alpine the handling of the tag (which is what you wanted to do reading your code) so the browser doesn't interfere. 
6) the other tags just update correctly in any case

**Reply by maxsite** [reply]
Your example is not working correctly. Press tab2 and a block will appear. Click again on tab 2 - it should hide, but remains open. This is wrong behavior.

Your first link is a 404 page.

**Reply by SimoTod** [reply]
It doesn't hide because in your code you set tab == 2. if you hide it, your component state won't match what the  browser shows. For example the span will say 2 but the second tab will be closed. I think i didn't understand fully what you wanted to do.

**Reply by maxsite** [reply]
No. Boolean html attributes don't work that way.
https://html.spec.whatwg.org/multipage/common-microsyntaxes.html#boolean-attribute

To hide a boolean attribute, alpina must remove it from the html-code.

```
<details open>
    <summary>tab1</summary>
    <div>text — open</div>
</details>

<details open="false">
    <summary>tab2</summary>
    <div>text - open</div>
</details>

<details open="open">
    <summary>tab3</summary>
    <div>text - open</div>
</details>

<details>
    <summary>tab4</summary>
    <div>text - close</div>
</details>
```

The Alpine 2 does it, the Alpine 3 doesn't.

**Reply by maxsite** [reply]
v2: https://github.com/alpinejs/alpine/blob/v2.8.2/README.md#x-bind —  see «x-bind for boolean attributes».
v3: https://alpinejs.dev/directives/bind — The behavior is not described.

**Comment by maxsite** [comment, marked as answer]
Strange behavior, but this is how it works.

```
     <div x-data="{tab: 1}">
        <details :open="tab == 1" @click.debounce.1ms="tab = 1">
            <summary>tab1</summary>
            <div>text</div>
        </details>

        <details :open="tab == 2" @click.debounce.1ms="tab = 2">
            <summary>tab2</summary>
            <div>text</div>
        </details>

        <details :open="tab == 3" @click.debounce.1ms="tab = 3">
            <summary>tab3</summary>
            <div>text</div>
        </details>
        
       <hr>
        <div x-text="tab"></div> <!-- for control -->
    </div>
```
